# Lecture 2: Summary Statistics and Boulder Weather Data 
***

In this notebook we'll: 
- Use Pandas to compute summary statistics on Boulder weather data 
- Figure out how summary statistics like mean and standard deviation change under transformations of the data


First, as always, we'll load Numpy and Pandas using their common aliases, np and pd. 

In [1]:
import numpy as np 
import pandas as pd

The data we'll explore in this notebook concerns temperatures and other weather observations in Boulder County over the month of July 2017.  The data was obtained from the National Oceanic and Atmospheric Administration's [Climate.gov](https://www.climate.gov/) website.  You can find and download loads of climate-related data from NOAA [here](https://www.climate.gov/maps-data/datasets).   

The data is stored in a .csv file called clean_boulder_weather.csv.  If you've updated your local course repository then you can read the data into pandas from the given local path.  If you have not (e.g. if you're using Azure) then you can load the data into pandas by passing in the url to the raw data file on GitHub.  

In [2]:
# Two different paths to the data 
local_path = 'data/clean_boulder_weather.csv'
web_path   = 'https://raw.githubusercontent.com/chrisketelsen/csci3022/master/inclass-notebooks/data/clean_boulder_weather.csv'

# Select the path that works for you 
file_path = web_path 

# Load the data into a DataFrame 
df = pd.read_csv(file_path)

Take a look at the first 50 or so rows of the DataFrame using the head( ) method. 

In [3]:
df.head(50)

STATION                    NAME        DATE  PRCP  TMAX  TMIN
0   USW00094075     BOULDER 14 W, CO US  2017-07-01  0.00  68.0  31.0
1   USW00094075     BOULDER 14 W, CO US  2017-07-02  0.00  73.0  35.0
2   USW00094075     BOULDER 14 W, CO US  2017-07-03  0.00  68.0  46.0
3   USW00094075     BOULDER 14 W, CO US  2017-07-04  0.05  68.0  43.0
4   USW00094075     BOULDER 14 W, CO US  2017-07-05  0.01  73.0  40.0
5   USW00094075     BOULDER 14 W, CO US  2017-07-06  0.00  76.0  48.0
6   USW00094075     BOULDER 14 W, CO US  2017-07-07  0.02  74.0  43.0
7   USW00094075     BOULDER 14 W, CO US  2017-07-08  0.00  65.0  44.0
8   USW00094075     BOULDER 14 W, CO US  2017-07-09  0.01  73.0  39.0
9   USW00094075     BOULDER 14 W, CO US  2017-07-10  0.01  75.0  44.0
10  USW00094075     BOULDER 14 W, CO US  2017-07-11  0.01  71.0  52.0
11  USW00094075     BOULDER 14 W, CO US  2017-07-12  0.10  54.0  43.0
12  USW00094075     BOULDER 14 W, CO US  2017-07-13  0.00  67.0  41.0
13  USW00094075     BOULDER 14 W, CO US  2017-07-14  0.00  67.0  39.0
14  USW00094075     BOULDER 14 W, CO US  2017-07-15  0.00  65.0  40.0
15  USW00094075     BOULDER 14 W, CO US  2017-07-16  0.05  68.0  38.0
16  USW00094075     BOULDER 14 W, CO US  2017-07-17  0.01  72.0  47.0
17  USW00094075     BOULDER 14 W, CO US  2017-07-18  0.00  71.0  47.0
18  USW00094075     BOULDER 14 W, CO US  2017-07-19  0.03  80.0  45.0
19  USW00094075     BOULDER 14 W, CO US  2017-07-20  0.00  74.0  46.0
20  USW00094075     BOULDER 14 W, CO US  2017-07-21  0.16  74.0  48.0
21  USW00094075     BOULDER 14 W, CO US  2017-07-22  0.11  66.0  46.0
22  USW00094075     BOULDER 14 W, CO US  2017-07-23  0.56  69.0  43.0
23  USW00094075     BOULDER 14 W, CO US  2017-07-24  0.00  72.0  45.0
24  USW00094075     BOULDER 14 W, CO US  2017-07-25  0.24  69.0  47.0
25  USW00094075     BOULDER 14 W, CO US  2017-07-26  0.28  62.0  44.0
26  USW00094075     BOULDER 14 W, CO US  2017-07-27  0.02  65.0  44.0
27  USW00094075     BOULDER 14 W, CO US  2017-07-28  0.12  69.0  45.0
28  USW00094075     BOULDER 14 W, CO US  2017-07-29  0.14  59.0  47.0
29  USW00094075     BOULDER 14 W, CO US  2017-07-30  0.48  61.0  41.0
30  USW00094075     BOULDER 14 W, CO US  2017-07-31  0.00  65.0  37.0
31  USC00053629  GROSS RESERVOIR, CO US  2017-07-01  0.00  63.0  40.0
32  USC00053629  GROSS RESERVOIR, CO US  2017-07-02  0.00  76.0  45.0
33  USC00053629  GROSS RESERVOIR, CO US  2017-07-03  0.00  83.0  50.0
34  USC00053629  GROSS RESERVOIR, CO US  2017-07-04  0.00  82.0  52.0
35  USC00053629  GROSS RESERVOIR, CO US  2017-07-05  0.05  82.0  47.0
36  USC00053629  GROSS RESERVOIR, CO US  2017-07-06  0.00  86.0  50.0
37  USC00053629  GROSS RESERVOIR, CO US  2017-07-07  0.00  86.0  50.0
38  USC00053629  GROSS RESERVOIR, CO US  2017-07-08  0.00  77.0  48.0
39  USC00053629  GROSS RESERVOIR, CO US  2017-07-09  0.03  78.0  48.0
40  USC00053629  GROSS RESERVOIR, CO US  2017-07-10  0.00  84.0  50.0
41  USC00053629  GROSS RESERVOIR, CO US  2017-07-11  0.00  82.0  55.0
42  USC00053629  GROSS RESERVOIR, CO US  2017-07-12  0.00  82.0  61.0
43  USC00053629  GROSS RESERVOIR, CO US  2017-07-13  0.07  68.0  46.0
44  USC00053629  GROSS RESERVOIR, CO US  2017-07-14  0.00  77.0  48.0
45  USC00053629  GROSS RESERVOIR, CO US  2017-07-15  0.00  78.0  52.0
46  USC00053629  GROSS RESERVOIR, CO US  2017-07-16  0.00  77.0  48.0
47  USC00053629  GROSS RESERVOIR, CO US  2017-07-17  0.00  78.0  50.0
48  USC00053629  GROSS RESERVOIR, CO US  2017-07-18  0.00  83.0  53.0
49  USC00053629  GROSS RESERVOIR, CO US  2017-07-19  0.00  83.0  54.0

From this you should see that each row in the DataFrame refers to a particular weather station / date combination.  The columns of the DataFrame are as follows: 

- **STATION**: The unique identification code for each weather station 
- **NAME**: The location / name of the weather station 
- **DATE**: The date of the observation 
- **PRCP**: The precipitation (in inches)
- **TMAX**: The daily maximum temperature (in Fahrenheit)
- **TMIN**: The daily minimum temperature (in Fahrenheit)

From the printed DataFrame above you can see that we actually have data from multiple weather stations.  To see how many, we can pass the **NAME** column (or the **STATION** column) into Python's set function. 

In [4]:
set(df["NAME"])

{'BOULDER 14 W, CO US',
 'BOULDER, CO US',
 'GROSS RESERVOIR, CO US',
 'NIWOT, CO US',
 'NORTHGLENN, CO US',
 'RALSTON RESERVOIR, CO US',
 'SUGARLOAF COLORADO, CO US'}

It looks like we have data from seven different weather stations.  For consistency, let's reduce the data to just the reports from the weather station in Niwot.  

### Exercise 1
***
Extract the rows of the DataFrame concerned with the Niwot weather station.  Store this data in a new DataFrame called dfNiwot. 

In [5]:
dfNiwot = df.loc[df["NAME"] == 'NIWOT, CO US'].copy()
dfNiwot.head() 

STATION          NAME        DATE  PRCP  TMAX  TMIN
93  USS0005J42S  NIWOT, CO US  2017-07-01   0.0  69.0  32.0
94  USS0005J42S  NIWOT, CO US  2017-07-02   0.0  73.0  37.0
95  USS0005J42S  NIWOT, CO US  2017-07-03   0.0  68.0  47.0
96  USS0005J42S  NIWOT, CO US  2017-07-04   0.1  70.0  41.0
97  USS0005J42S  NIWOT, CO US  2017-07-05   0.0  74.0  40.0

### Exercise 2  
***

Pandas (and Numpy) have canned functions that compute each of the summary statistics discussed in lecture.  We'll use the .mean( ) function as an example.  All of these functions can be called either on a Pandas Series (i.e. a column of a DataFrame) or on an entire DataFrame at one time.  

For instance, the sample mean of the maximum daily temperature is given by 

In [6]:
dfNiwot["TMAX"].mean()

69.83870967741936

Let's see what happens if we call .mean( ) on the entire DataFrame. 

In [7]:
dfNiwot.mean()

PRCP     0.061290
TMAX    69.838710
TMIN    43.548387
dtype: float64

In this case, Pandas returned a Series with the means of all of the **numerical** data in the DataFrame. 

The functions for the other summary statistics are as follows: 

\begin{array}{l|l}
\textrm{Function} & \textrm{Statistics} \\
\hline
\textrm{.var()} & \textrm{variance} \\
\textrm{.std()} & \textrm{standard deviation} \\
\textrm{.min()} & \textrm{minimum value} \\
\textrm{.max()} & \textrm{maximum value} \\
\textrm{.median()} & \textrm{value} \\
\textrm{.quantile(q)} & \textrm{quantile, where q is the desired percentage as a decimal} \\
\end{array}

Your job is to use these functions to compute the 5-number summary for the maximum daily temperature. 

In [8]:
minval = dfNiwot["TMAX"].min()
maxval = dfNiwot["TMAX"].max()
Q1 = dfNiwot["TMAX"].quantile(.25)
Q2 = dfNiwot["TMAX"].quantile(.50) # Could also do dfNiwot["TMAX"].median()
Q3 = dfNiwot["TMAX"].quantile(.75)
print("5-Number Summary: {:.2f}    {:.2f}    {:.2f}    {:.2f}    {:.2f}".format(minval, Q1, Q2, Q3, maxval))

5-Number Summary: 54.00    66.50    70.00    74.00    80.00


### Exercise 3 
***
It turns out that Pandas has a nice function called .describe( ) that will compute all of the standard summary statistics for you.  You can apply it either to a Pandas Series or to an entire DataFrame.  

Run the .describe( ) function on the **TMAX** column of your DataFrame, and check that the results agree with your computations from Exercise 2. 

In [9]:
dfNiwot["TMAX"].describe()

count    31.000000
mean     69.838710
std       5.621962
min      54.000000
25%      66.500000
50%      70.000000
75%      74.000000
max      80.000000
Name: TMAX, dtype: float64

### Exercise 4 
***
In this exercise we'll explore how the mean and the standard deviation change when we perform basic transformations on the data.  In particular, we're interested in what happens if we 

1. Add or subtract some value from every entry in the data set 
1. Multiply every entry in the data set by some value 

We know from above that the mean and standard deviation of the Niwot **TMAX** value are 69.83871 and 5.621962.  Experiment by adding and multiplying nice integer values with the **TMAX** column and then recomputing the statistics.  From your observations, can you guess how the mean and std dev change under these transformations? 

In [10]:
print((dfNiwot["TMAX"]+3).mean())
print((dfNiwot["TMAX"]+3).std())
print((dfNiwot["TMAX"]*3).mean())
print((dfNiwot["TMAX"]*3).std())

72.83870967741936
5.621961544950591
209.51612903225808
16.865884634851767


See if you can prove that your guess works in general mathematically using the formulas for the two statistics: 

$$
\bar{x} = \frac{1}{n} \displaystyle\sum_{k=1}^n x_k \quad \quad \textrm{and} \quad \quad s = \sqrt{\frac{1}{n-1} \sum_{k=1}^n \left( x_k - \bar{x}\right)^2} 
$$

**Solution**: 

**The Mean with Addition**: It looks like when we add a constant to each observation the constant also gets added to the mean.  We can show this in general as follows.  Let $y_k = x_k + a$ be the shifted observations.  We then have  


$$
\bar{y} = \frac{1}{n} \sum_{k=1}^n y_k \quad = \quad \frac{1}{n} \sum_{k=1}^n (x_k + a) \quad = \quad 
\frac{1}{n} \sum_{k=1}^n x_k  + \frac{1}{n} \sum_{k=1}^n a 
\quad = \quad \bar{x}  + \frac{1}{n} \cdot an 
\quad = \quad \bar{x}  + a 
$$

**The Std Dev with Addition**: On the other hand, it looks like the standard deviation stays the same when we add a constant to each observation.  This should make intuitive sense because the std dev is a measure of the spread of the data, and by adding a constant to each observation we're just shifting things down the number line.  Let's see if we can use the formula for std dev to confirm this mathematically. 

$$
\sqrt{\frac{1}{n-1}\sum_{k=1}^n \left( y_k - \bar{y} \right)^2 } \quad = \quad 
\sqrt{\frac{1}{n-1}\sum_{k=1}^n \left[ (x_k + a) - (\bar{x}+a) \right]^2 } \quad = \quad 
\sqrt{\frac{1}{n-1}\sum_{k=1}^n \left( x_k - \bar{x} \right)^2 } 
$$

And so we see that the standard deviation of both the $x$'s and the  $y$'s are the same. 

**The Mean with Multiplication**: It looks like when we multiply each observation by a constant the mean also gets multiplied by the constant.  We can show this in general as follows.  Let $z_k = b \cdot x_k$ be the multiplied observations.  We then have  


$$
\bar{z} = \frac{1}{n} \sum_{k=1}^n z_k \quad = \quad \frac{1}{n} \sum_{k=1}^n b \cdot x_k  \quad = \quad 
b \cdot \frac{1}{n} \sum_{k=1}^n x_k  
\quad = \quad b\cdot \bar{x}  
$$

**The Std Dev with Addition**: It also looks like the std dev gets multiplied by the constant.  Let's see 

$$
\sqrt{\frac{1}{n-1}\sum_{k=1}^n \left( z_k - \bar{z} \right)^2 } \quad = \quad 
\sqrt{\frac{1}{n-1}\sum_{k=1}^n \left( b\cdot x_k - b\cdot \bar{x} \right)^2 } \quad = \quad 
\sqrt{\frac{1}{n-1}\sum_{k=1}^n b^2 \cdot \left( x_k - \bar{x} \right)^2 } \quad = \quad 
b\cdot \sqrt{\frac{1}{n-1}\sum_{k=1}^n  \left( x_k - \bar{x} \right)^2 } 
$$

Confirmed! 

### Exercise 5 
***
OK, let's apply a common transformation to the **TMAX** and **TMIN** columns by converting the temperatures from Fahrenheit to Celsius.  Remember that the transformation is given by 

$$
\textrm{CELSIUS} = \frac{5}{9} (\textrm{FAHRENHEIT}-32) 
$$

First, use the Fahrenheit data in columns **TMAX** and **TMIN** to create Celsius columns in the Niwot DataFrame called **TMAX-C** and **TMIN-C**.

In [11]:
dfNiwot.loc[:,"TMAX-C"] = (5/9)*(dfNiwot.loc[:,"TMAX"]-32)
dfNiwot.loc[:,"TMIN-C"] = (5/9)*(dfNiwot.loc[:,"TMIN"]-32)

In [12]:
dfNiwot.head()

STATION          NAME        DATE  PRCP  TMAX  TMIN     TMAX-C  \
93  USS0005J42S  NIWOT, CO US  2017-07-01   0.0  69.0  32.0  20.555556   
94  USS0005J42S  NIWOT, CO US  2017-07-02   0.0  73.0  37.0  22.777778   
95  USS0005J42S  NIWOT, CO US  2017-07-03   0.0  68.0  47.0  20.000000   
96  USS0005J42S  NIWOT, CO US  2017-07-04   0.1  70.0  41.0  21.111111   
97  USS0005J42S  NIWOT, CO US  2017-07-05   0.0  74.0  40.0  23.333333   

      TMIN-C  
93  0.000000  
94  2.777778  
95  8.333333  
96  5.000000  
97  4.444444


Based on the stuff we proved in **Exercise 4**, what do you expect the mean and the standard deviation of the daily maximum temperature to be in Celsius? 

**Solution**: 

Let $\bar{x}^{min}$ and $\bar{x}^{max}$ represent the mean of **TMAX** and **TMIN** in Fahrenheit, respectively, then we expect the means of the associated values in Celsius to be 

$$
\bar{y}^{min} = \frac{5}{9}\left( \bar{x}^{min} - 32 \right) = \frac{5}{9}\left(43.548 -32 \right) = 6.416
\quad \textrm{and} \quad 
\bar{y}^{max} = \frac{5}{9}\left( \bar{x}^{max} - 32 \right) = \frac{5}{9}\left(69.839 -32 \right) = 21.022
$$

Once you've made your guess, see if you're right by applying the .mean( ) method to **TMAX-C** and **TMIN-C**. 

In [13]:
ybar_min = dfNiwot["TMIN-C"].mean()
ybar_max = dfNiwot["TMAX-C"].mean()
print("Mean Min Temp in Celsius = {:.3f}".format(ybar_min))
print("Mean Max Temp in Celsius = {:.3f}".format(ybar_max))

Mean Min Temp in Celsius = 6.416
Mean Max Temp in Celsius = 21.022


### Exercise 6 
***

Compute the daily temperature range (max minus min) in Fahrenheit for each row in the Niwot DataFrame and store it in a column called **TDIFF**.  Then answer these questions.  

- What is the mean temperature difference over the month of July? 
- What is the difference between the means of the max and min daily temperatures? 
- Do you see a relationship between these two quantities?  If so, can you prove that it's always the case for mean difference and difference of means? 

In [14]:
dfNiwot.loc[:,"TDIFF"] = dfNiwot.loc[:,"TMAX"] - dfNiwot.loc[:,"TMIN"]

In [15]:
dfNiwot.head()

STATION          NAME        DATE  PRCP  TMAX  TMIN     TMAX-C  \
93  USS0005J42S  NIWOT, CO US  2017-07-01   0.0  69.0  32.0  20.555556   
94  USS0005J42S  NIWOT, CO US  2017-07-02   0.0  73.0  37.0  22.777778   
95  USS0005J42S  NIWOT, CO US  2017-07-03   0.0  68.0  47.0  20.000000   
96  USS0005J42S  NIWOT, CO US  2017-07-04   0.1  70.0  41.0  21.111111   
97  USS0005J42S  NIWOT, CO US  2017-07-05   0.0  74.0  40.0  23.333333   

      TMIN-C  TDIFF  
93  0.000000   37.0  
94  2.777778   36.0  
95  8.333333   21.0  
96  5.000000   29.0  
97  4.444444   34.0

To compute the mean temperature difference we just need to compute the mean of the **TDIFF** column we just created

In [16]:
mean_diff = dfNiwot["TDIFF"].mean()
print("Mean Temp Diff = {:.3f}".format(mean_diff))

Mean Temp Diff = 26.290


We compute the difference of the max and min temperature means as follows 

In [17]:
diff_of_means = dfNiwot["TMAX"].mean() - dfNiwot["TMIN"].mean()
print("Diff of Mean Temps = {:.3f}".format(diff_of_means))

Diff of Mean Temps = 26.290


It looks like they're the same! Can we show mathematically that this is true in general? 

**Solution**: 

Suppose we have $n$ observations of two variables $x_k$ and $y_k$ for $k=1,\ldots,n$, and define $z_k = y_k - x_k$.  We want to show that the 

$$
\bar{z} = \bar{y} - \bar{x}.
$$

We'll do this using the definition of the sample mean.  We have 

\begin{eqnarray}
\bar{z} &=& \frac{1}{n}\sum_{k=1}^n z_k  \\
&=& \frac{1}{n}\sum_{k=1}^n (y_k - x_k)   \\
&=& \frac{1}{n}\sum_{k=1}^n y_k - \frac{1}{n}\sum_{k=1}^n x_k  \\ 
&=& \bar{y} - \bar{x}
\end{eqnarray}

**Note** that this little fact is only relevant when we have the same number of observations for $x$ and $y$. 